docs/CIP_6_Pipelines_Registry.md

```diff
@@ -0,0 +1,309 @@
+
+#### Pipeline Step Registration - The Registry
+
+For example, if you wanted to create a pipeline that chunked a document, you could do something like this:
+
+First, register The "Chunk" pipeline step with the name "Chunk" as follows:
+
+```python
+@pipeline_step(name="Chunk")
+def chunk(data: str) -> List[str]:
+    # Chunk the data
+    return chunked_data
+```
+ <!-- NOTE: ALTERNATIVE CONSIDERED = DO A CLASS BASED APPROACH, WHERE THE CLASS IS THE PIPELINE STEP -->
+
+Then you can create a pipeline that chunks a document as follows:
+```python
+to_add = Chunk(data)
+```
+
+Calling Chunk() does not actually execute the pipeline step, instead it returns a reference to the pipeline step. This reference is then passed to the add step, which is responsible for executing the pipeline step. This is done to allow for the pipeline step to be executed on the server-side. A pipeline can always be executed locally by calling the run() method on the pipeline.
+
+The assumption then, is that you provide the server with the relevant pipeline steps to execute. The exact mechanism for this is later defined. For local execution, the pipeline steps are executed on the client-side.
+
+#### Pipeline Step Composition
+
+Pipeline steps can be composed, and return a reference to the composed pipeline step. For example, if you wanted to create a pipeline that chunked a document and then embedded the chunks, you could do something like this:
+
+```python
+to_add = Embed(Chunk(data))
+```
+
+Or perhaps you want to chunk, embed and then quantize the embeddings:
+
+```python
+to_add = Quantize(Embed(Chunk(data)))
+```
+
+NOTE: This syntax could be confusing as the order of operations is reversed. We could consider alternative designs.
+
+#### Default Pipeline and Pipeline Persistence
+
+It is important to preserve the existing behavior of our API where users can simply pass text to the add step and have it be embedded. To do this, we will create a default pipeline that is used when no pipeline is provided. This default pipeline will simply embed the text. This default pipeline will be persisted with the collection, and will be used when no pipeline is provided. This means that users can create a collection with a default pipeline, and then query the collection without having to provide the pipeline. This is a significant improvement over the current system, where users must remember which embedding function they used when they created the collection.
+
+Pipelines can be persisted by serializing the pipeline steps and storing them in the metadata of the collection. This means that when a collection is loaded, the pipeline steps can be deserialized and used to recreate the pipeline. This is a significant improvement over the current system, where users must remember which embedding function they used when they created the collection.
```

Comment:
+1 for DLS/portable format. One caveat is when we get pipeline function sprawl from community contributions. Lots of projects split functions into core and contrib.